verifier: Fetch VCEK cert from KDS instead of bailing

AdithyaKrishnan opened the pull request:
Fetch the VCEK cert from the KDS if it is absent in the cert chain instead of just printing a bail statement stating that the VCEK is not found.

fitzthum commented on deps/verifier/Cargo.toml:
```diff
@@ -49,6 +49,7 @@ strum.workspace = true
 veraison-apiclient = { git = "https://github.com/chendave/rust-apiclient", branch = "token", optional = true }
 ear = { git = "https://github.com/veraison/rust-ear", rev = "43f7f480d09ea2ebc03137af8fbcd70fe3df3468", optional = true }
 x509-parser = { version = "0.14.0", optional = true }
+reqwest = { version="0.12.9", features = ["blocking"] }
```
I you are already specifying the needed version of `reqwest` in the global Cargo file, you can just say `request = workspace.true` here.

AdithyaKrishnan replied:
Changed. If appropriate, please mark this comment as Resolved.

fitzthum replied:
See Ding's suggestions about the Cargo file as well.

fitzthum commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -84,8 +85,10 @@ impl Verifier for Snp {
 
-        let Some(cert_chain) = cert_chain else {
-            bail!("Cert chain is unset");
+        let cert_chain = if let Some(chain) = cert_chain{
```
So the cert chain is type `Option<Vec<Cert>>` or something. This accounts for cases where the option is none, but what if the option is there but the vector is empty?

Ideally we should get the vcek even if there is a non-empty cert chain as long as the vcek/vlek isn't part of it.

AdithyaKrishnan replied:
Could this be part of the future follow-up PR?

fitzthum replied:
We can leave check the cert types to a follow-up, but please add a case to this statement to catch an empty vector.

AdithyaKrishnan replied:
Added the empty vector check

fitzthum commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -307,6 +310,38 @@ fn get_common_name(cert: &x509::X509) -> Result<String> {
 }
 
+// Function to request vcek from KDS. Return vcek in der format.
+fn request_endorsement_key_kds(
```
nit: `request_endorsement_key_kds` -> `get_endorsement__key_from_kds`

AdithyaKrishnan replied:
Per Claudio's suggestion, I've renamed it to fetch_vcek_from_kds().  If appropriate, please mark this comment as Resolved.

fitzthum commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -307,6 +310,38 @@ fn get_common_name(cert: &x509::X509) -> Result<String> {
+) -> Result<Vec<CertTableEntry>, anyhow::Error> {
+    // KDS URL parameters
+    const KDS_CERT_SITE: &str = "https://kdsintf.amd.com";
+    const KDS_VCEK: &str = "/vcek/v1";
```
might be better to put these consts at the top of the file rather than inside this function

AdithyaKrishnan replied:
Fixed. If appropriate, please mark this comment as Resolved.

fitzthum commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -307,6 +310,38 @@ fn get_common_name(cert: &x509::X509) -> Result<String> {
+        StatusCode::OK => {
+            let vcek_rsp_bytes: Vec<u8> =
+                vcek_rsp.bytes().context("Unable to parse VCEK")?.to_vec();
+            let key= CertTableEntry{cert_type: CertType::VCEK, data: vcek_rsp_bytes};
```
nit: `key=` -> `key =`. Also you probably want some spaces around the `{`. I'm surprised this survived `cargo fmt` check.

AdithyaKrishnan replied:
If appropriate, please mark this comment as Resolved.

mkulke replied:
there are more formatting/lint issues. we do have a problem in the build pipeline, the verifier crate is not linted/formatted.

mkulke replied:
can you run

```bash
cargo fmt
cargo clippy -p verifier -- -D warnings
```

on your code in the mean time and fix the warnings?

AdithyaKrishnan replied:
Ran both the commands and the latest commit reflects the same

cclaudio commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -307,6 +310,38 @@ fn get_common_name(cert: &x509::X509) -> Result<String> {
+// Function to request vcek from KDS. Return vcek in der format.
+fn request_endorsement_key_kds(
+    att_report: AttestationReport,
+) -> Result<Vec<CertTableEntry>, anyhow::Error> {
```
The function is specific to download the VCEK, so I think it could be rename to something more specific such as `fetch_vcek_from_kds()`

cclaudio replied:
Additionally, I think the function would be more self-contained if it returns `Result<Vec<u8>>`

AdithyaKrishnan replied:
Changed. If appropriate, please mark this comment as Resolved.

AdithyaKrishnan replied:
> Additionally, I think the function would be more self-contained if it returns `Result<Vec<u8>>`

Tried this and got a type mismatch error

fitzthum replied:
Btw I don't think you need `anyhow::Error` in in this return type.

AdithyaKrishnan replied:
Fixed

cclaudio reviewed (commented):
Just a few minor comments.

fitzthum reviewed (commented):
Nice PR. Looks like this will address #456 which is actually a duplicate of #363 

One thing to think about is that this could result in quite a few requests being made to the KDS. We should probably implement some kind of caching mechanism so that the verifier doesn't have to get the `vcek` for the same node again and again. We can probably do this in a follow-up. 

Also, we might want to add `vlek` support at some point although I suspect this will require specifying a few more config params.

cclaudio commented on Cargo.lock:
It seems your Cargo.lock has more changes than required. You may need to restore it and apply the `request = worspace.true` change that Tobin suggested

AdithyaKrishnan replied:
If appropriate, please mark this comment as Resolved.

Xynnn007 commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -482,4 +518,4 @@ mod tests {
         verify_report_signature(&attestation_report, &cert_chain, vendor_certs).unwrap_err();
     }
-}
+}
```
this empty line can be reserved

AdithyaKrishnan replied:
Fixed

Xynnn007 commented on Cargo.toml:
```diff
@@ -36,7 +36,7 @@ jsonwebtoken = { version = "9", default-features = false }
 prost = "0.12"
 regorus = { version = "0.1.5", default-features = false, features = ["regex", "base64", "time"] }
-reqwest = { version = "0.12", default-features = false, features = ["default-tls"] }
+reqwest = { version = "0.12", default-features = false, features = ["default-tls", "blocking"] }
```
I noticed that the `reqwest` is used in `fetch_vcek_from_kds`. It is suggested to use `reqwest::get` (the async version) rather than blocking version, because the blocking version would hang on network I/O during which the process would wait.

AdithyaKrishnan replied:
I'm able to resolve most comments except this. Could we please have a conversation on slack?

Xynnn007 commented on deps/verifier/Cargo.toml:
```diff
@@ -49,6 +49,7 @@ strum.workspace = true
 veraison-apiclient = { git = "https://github.com/chendave/rust-apiclient", branch = "token", optional = true }
 ear = { git = "https://github.com/veraison/rust-ear", rev = "43f7f480d09ea2ebc03137af8fbcd70fe3df3468", optional = true }
 x509-parser = { version = "0.14.0", optional = true }
+reqwest.workspace = true
```
It is good to mark this dependency to feature `snp-verifier`. s.t.
```suggestion
reqwest = { optional = true, workspace = true }
```

and add `"request"` to previous `[features]` section
```toml
...
[features]
snp-verifier = [ "asn1-rs", "openssl", "sev", "x509-parser", "reqwest" ]
...
```

fitzthum commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -307,6 +314,35 @@ fn get_common_name(cert: &x509::X509) -> Result<String> {
+            let key = CertTableEntry { cert_type: CertType::VCEK, data: vcek_rsp_bytes };
+            Ok(vec![key])
+        }
+        status => Err(anyhow::anyhow!("Unable to fetch VCEK from URL: {status:?}")),
```
`anyhow::anyhow!` -> `anyhow!`

AdithyaKrishnan replied:
fixed

fitzthum commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -84,8 +92,14 @@ impl Verifier for Snp {
+                chain
+            }
+        } else {
+            fetch_vcek_from_kds(report)?
```
I think you can simplify this by doing something like 

```
let cert_chain = match cert_chain {
    Some(chain) if !chain.is_empty() => chain,
    _ => fetch_vcek_from_kds(report)?
}
```
Or something like that

AdithyaKrishnan replied:
Fixed!

Xynnn007 commented on deps/verifier/src/snp/mod.rs:
```diff
@@ -307,6 +313,41 @@ fn get_common_name(cert: &x509::X509) -> Result<String> {
     Ok(e.data().as_utf8()?.to_string())
 }
 
+// Function to request vcek from KDS asynchronously. Return vcek in der format.
```
We can change `//` to `///`. This would automatically [generate a document](https://doc.rust-lang.org/rust-by-example/meta/doc.html#doc-comments) about the function when we publish the crate to `crates.io`.

AdithyaKrishnan replied:
Do you want me to add description to all functions in the file?

Xynnn007 reviewed (approved):
Thanks @AdithyaKrishnan ! LGTM

AdithyaKrishnan commented:
> Thanks @AdithyaKrishnan ! LGTM. Please see @fitzthum and @cclaudio 's comments.

@fitzthum and @cclaudio  - I believe I've fixed all your concerns. Please let me know if there are any more to be resolved. Further, per Ding's suggestion, I've added a third commit that adds on to the documentation by providing descriptions of all functions.
Hoping this PR can be approved and merged at the earliest.

fitzthum reviewed (approved):
LGTM

cclaudio reviewed (approved):
LGTM.  Thanks @AdithyaKrishnan
